bmp180/bmp180.go

```diff
@@ -0,0 +1,191 @@
+// https://cdn-shop.adafruit.com/datasheets/BST-BMP180-DS000-09.pdf
+//
+package bmp180
+
+import (
+	"time"
+
+	"github.com/aykevl/tinygo/src/machine"
+)
+
+const (
+	// ULTRALOWPOWER is the lowest oversampling mode of the pressure measurement.
+	ULTRALOWPOWER OversamplingMode = iota
+	// BSTANDARD is the standard oversampling mode of the pressure measurement.
+	STANDARD
+	// HIGHRESOLUTION is a high oversampling mode of the pressure measurement.
+	HIGHRESOLUTION
+	// ULTRAHIGHRESOLUTION is the highest oversampling mode of the pressure measurement.
+	ULTRAHIGHRESOLUTION
+)
+
+// BMP180OversamplingMode is the oversampling ratio of the pressure measurement.
+type OversamplingMode uint
+
+// calibrationCoefficients reads at startup and stores the calibration coefficients
+type calibrationCoefficients struct {
+	ac1 int16
+	ac2 int16
+	ac3 int16
+	ac4 uint16
+	ac5 uint16
+	ac6 uint16
+	b1  int16
+	b2  int16
+	mb  int16
+	mc  int16
+	md  int16
+}
+
+// Device wraps an I2C connection to a BMP180 device.
+type Device struct {
+	bus                     machine.I2C
+	mode                    OversamplingMode
+	calibrationCoefficients calibrationCoefficients
+}
+
+// New creates a new BMP180 connection. The I2C bus must already be
+// configured.
+//
+// This function only creates the Device object, it does not touch the device.
+func New(bus machine.I2C) Device {
+	return Device{
+		bus:  bus,
+		mode: ULTRAHIGHRESOLUTION,
+	}
+}
+
+// Connected returns whether a BMP180 has been found.
+// It does a "who am I" request and checks the response.
+func (d *Device) Connected() bool {
+	data := []byte{0}
+	d.bus.ReadRegister(Address, WHO_AM_I, data)
+	return data[0] == CHIP_ID
+}
+
+// Configure sets up the device for communication and
+// read the calibration coefficientes.
+func (d *Device) Configure() {
+	data := make([]byte, 22)
+	err := d.bus.ReadRegister(Address, AC1_MSB, data)
+	if err != nil {
+		return
+	}
+	d.calibrationCoefficients.ac1 = readInt(data[0], data[1])
+	d.calibrationCoefficients.ac2 = readInt(data[2], data[3])
+	d.calibrationCoefficients.ac3 = readInt(data[4], data[5])
+	d.calibrationCoefficients.ac4 = readUint(data[6], data[7])
+	d.calibrationCoefficients.ac5 = readUint(data[8], data[9])
+	d.calibrationCoefficients.ac6 = readUint(data[10], data[11])
+	d.calibrationCoefficients.b1 = readInt(data[12], data[13])
+	d.calibrationCoefficients.b2 = readInt(data[14], data[15])
+	d.calibrationCoefficients.mb = readInt(data[16], data[17])
+	d.calibrationCoefficients.mc = readInt(data[18], data[19])
+	d.calibrationCoefficients.md = readInt(data[20], data[21])
+}
+
+// Temperature returns the temperature in celsius degrees (ºC)
+func (d *Device) Temperature() (temperature float32, err error) {
+	rawTemp, err := d.rawTemp()
+	if err != nil {
+		return
+	}
+	b5 := d.calculateB5(rawTemp)
+	t := (b5 + 8) >> 4
+	return float32(t) / 10, nil
+}
+
+// Pressure returns the pressure in pascals (Pa)
+func (d *Device) Pressure() (pressure float32, err error) {
```

Comment:
I know this is an ugly calculation, perhaps a link back to the datasheet showing why that is the case would be helpful for anyone trying to interpret these "magic numbers".

What do you think?

Reply:
The datasheet (already linked on top) did not have deeplinks, I had no way to link directly to the algo/calculations. But you are right, I should add a note stating where those numbers come from at least.

The official implementation (recommended on the datasheet) is ... hard to look at: https://github.com/BoschSensortec/BMP180_driver/blob/master/bmp180.c#L252